library/ssl_tls.c

```diff
@@ -5818,6 +5867,71 @@ void mbedtls_ssl_handshake_free( mbedtls_ssl_context *ssl )
 
     mbedtls_platform_zeroize( handshake,
                               sizeof( mbedtls_ssl_handshake_params ) );
+
+#if defined(MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH)
+    /* If the buffers are too big - reallocate */
+    {
+        int modified = 0;
+        unsigned char* resized_buffer = NULL;
+        uint32_t buf_len = mbedtls_ssl_get_input_buflen( ssl );
+        size_t written_in = 0;
+        size_t written_out = 0;
+        if( ssl->in_buf != NULL &&
+            ssl->in_buf_len > buf_len &&
+            ssl->in_left < buf_len )
+        {
+            resized_buffer = mbedtls_calloc( 1,  buf_len );
+            if( resized_buffer != NULL )
+            {
+                MBEDTLS_SSL_DEBUG_MSG( 2, ( "Reallocating in_buf to %d", buf_len ) );
+                written_in = ssl->in_msg - ssl->in_buf;
+                memcpy( resized_buffer, ssl->in_buf, buf_len );
+                mbedtls_platform_zeroize( ssl->in_buf, ssl->in_buf_len );
+                mbedtls_free( ssl->in_buf );
+
+                ssl->in_buf = resized_buffer;
```

Comment:
Does LCOV show we reached here in our unit testing?

Reply:
Good suggestion - while checking that I've found that i have decreased test coverage by doing the following:
- First, I had separate 'integration' tests doing every single feature + variable buffers and serialization + MFL + variable buffers, renegotiation + MFL + variable buffers.
- Then I figured that if we don't want to duplicate so many tests just to add the variable buffers, we could run the whole test suite with variable buffers turned on by default.
- I removed the additional tests doing every single feature + variable buffers, but by doing so, I also removed the `integration` tests - MFL + serialization + VB, MFL + renegotiation + VB.

I'll add these tests as a part of the last commit.